sync15-adapter/src/state_machine.rs

```diff
@@ -0,0 +1,599 @@
+                                // does Rust support multiple `if let Some(k) = x, let Some(l) = y { ... }
+                                let previous_engine = previous_global.engines.get(*name).unwrap();
+                                let engine = global.engines.get(*name).unwrap();
+                                if previous_engine.sync_id != engine.sync_id {
+                                    changes.push(EngineStateChange::Reset(name.to_string()));
+                                }
+                            }
+                            MetaGlobalEnginesAction::ChangeStates(changes)
+                        }
+                    }
+                    None => MetaGlobalEnginesAction::ResetAll,
+                };
+
+                match action {
+                    MetaGlobalEnginesAction::ResetAll => {
+                        scratchpad.set_global(Some(global));
+                        scratchpad.set_keys(None);
+                        scratchpad.add_engine_state_changes_from_keys(None);
+                        scratchpad.save();
+                        Ok(HasMetaGlobal { collections })
+                    }
+                    MetaGlobalEnginesAction::ChangeStates(changes) => {
+                        scratchpad.set_global(Some(global));
+                        for change in changes {
+                            scratchpad.add_engine_state_change(change);
+                        }
+                        scratchpad.save();
+                        Ok(HasMetaGlobal { collections })
+                    }
+                }
+            }
+
+            // Check if our locally cached `crypto/keys` collection is
+            // up-to-date.
+            HasMetaGlobal { collections } => {
+                let action = match scratchpad.keys() {
+                    Some(keys) => match &collections.get("crypto") {
+                        Some(modified) => {
+                            if keys.modified >= **modified {
+                                FetchAction::Skip
+                            } else {
+                                FetchAction::Fetch
+                            }
+                        }
+                        None => FetchAction::InvalidateThenFetch,
+                    },
+                    None => FetchAction::Fetch,
+                };
+                Ok(match action {
+                    // If `crypto/keys` is up-to-date, we're ready to go!
+                    FetchAction::Skip => Ready,
+                    // We need to fetch and cache new keys.
+                    FetchAction::Fetch => NeedsFreshCryptoKeys,
+                    // We need to invalidate our locally cached `crypto/keys`,
+                    // then try to fetch new keys, and reupload if fetching
+                    // fails.
+                    FetchAction::InvalidateThenFetch => {
+                        scratchpad.set_keys(None);
```

Comment:
Should there be a scratchpad.save() here? (and a few places below). Maybe saving should be implicit?